### PR TITLE
prop_find implementation

### PR DESCRIPTION
**Pull request by r-hackett**
This PR implements the prop_find function as discussed in issue #4970 and provides a test. The function f_prop_find is added in src/textprop.c and Test_prop_find in src/testdir/test_textprop.vim. I followed the documentation closely and the function should behave as intended, but to be specific I'll mention a few points with reference to the docs here.

> skipstart:    do not look for a match at the start position

'skipstart' is treated as a number, 0 being false (default) and 1 being true. If true and the text property at the start position matches the one being searched for (specified by 'id' or 'type'), then this property from its start to end line is ignored. This is the same searching forwards or backwards. 

> {direction} can be "f" for forward and "b" for backward.

This is treated as a string with the first character considered. The default is forward.

If a matching prop is found by prop_find, the first line (with the start flag set to 1) of the prop is returned in a dictionary along with an additional 'lnum' entry, otherwise an empty dict is returned.

The test function tests prop_find fairly thoroughly but may leave out some edge cases (although I've tested these myself by experimentation). More can of course be added.

I hope the code style is ok and I'm happy to document or explain the function better. Sorry for the terrible timing being new year but I thank you for your consideration and feedback.

### CODE REVIEW

**Comment by brammool** [comment]
Ryan Hackett wrote:

> This PR implements the prop_find function as discussed in issue #4970 and provides some tests. The function f_prop_find is added in src/textprop.c and Test_prop_find in src/testdir/test_textprop.vim. I followed the documentation closely and the function should behave as intended, but to be specific I&#39;ll mention a few points with reference to the docs here.
> 
> &gt; skipstart:    do not look for a match at the start position
> 
> &#39;skipstart&#39; is treated as a number, 0 being false (default) and 1 being true. If true and the text property at the start position matches the one being searched for (specified by &#39;id&#39; or &#39;type&#39;), then this property from its start to end line is ignored. This is the same searching forwards or backwards. 
> 
> &gt; {direction} can be &quot;f&quot; for forward and &quot;b&quot; for backward.
> 
> This is treated as a string with the first character considered. The default is forward.
> 
> If a matching prop is found by prop_find, the first line (with the start flag set to 1) of the prop is returned in a dictionary along with an additional &#39;lnum&#39; entry, otherwise an empty dict is returned.
> 
> The test function tests prop_find fairly thoroughly but may leave out some edge cases (although I&#39;ve tested these myself by experimentation). More can of course be added.
> 
> I hope the code style is ok and I&#39;m happy to document or explain the function better. Sorry for the terrible timing being new year but I thank you for your consideration and feedback. 

Thanks for working on this!

Filling the result dictionary can be shared with f_prop_list().
Move the common code to a function, prop_fill_dict() perhaps.

A few minor comments:
- Please fix the indents to use tabs consistently.
- There is one "if(" that should be "if (".

-- 
Mental Floss prevents moral decay!

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by r-hackett** [comment]
> Filling the result dictionary can be shared with f_prop_list().
> Move the common code to a function, prop_fill_dict() perhaps.

I created the prop_fill_dict function and made the necessary changes to prop_list and prop_find. It's static currently. Should it be visible outside of textprops.c?

> A few minor comments:
> - Please fix the indents to use tabs consistently.
> - There is one "if(" that should be "if (".

I forgot that I set vim up to indent with spaces but I've changed it to tabs and fixed each line. I also made a few changes to the indentation level to make the style more similar to the other functions.

> Thanks for working on this!

I'm very happy to contribute, thank you.

**Comment by brammool** [comment]
Ryan Hackett wrote:

> > Filling the result dictionary can be shared with f_prop_list().
> > Move the common code to a function, prop_fill_dict() perhaps.
> 
> I created the prop_fill_dict function and made necessary changes to
> prop_list and prop_find. It's static currently. Should it be visible
> outside of textprops.c?

Functions are static until they are used in multiple files.

> > A few minor comments:
> > - Please fix the indents to use tabs consistently.
> > - There is one "if(" that should be "if (".
> 
> Forgot I set vim up for spaces so I've done my best to use tabs in the
> added code. I did make a few changes to the indentation level to make
> the style more like the other functions, sorry if this makes the diff
> look a bit messy.

Somehow the indents are wrong now.  Re-indending should fix most of it.
The modeline has the relevant values.

-- 
The war between Emacs and Vi is over.  Vi has won with 3 to 1.
http://m.linuxjournal.com/files/linuxjournal.com/linuxjournal/articles/030/3044/3044s1.html

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by r-hackett** [comment]
That should fix the indents. A strange thing is that in the [diff](https://github.com/vim/vim/pull/5421/files#diff-be67b582e26932bd9217fba84b6c2e88) for textprop.c on github the variables at the top of prop_find are still not aligned, but in the [file viewer](https://github.com/vim/vim/blob/8dea31f4d55cdb3195b77a0d852b58f3a61883c5/src/textprop.c) (and in vim with now I've made sure modeline is set) they're aligned. Apologies for the number of commits here, should be more thorough.

File: src/textprop.c [hunk 1]
```diff
@@ -468,6 +468,24 @@ find_type_by_id(hashtab_T *ht, int id)
...
+    proptype_T *pt;
+
+    dict_add_number(dict, "col", prop->tp_col);
+    dict_add_number(dict, "length", prop->tp_len);
+    dict_add_number(dict, "id", prop->tp_id);
+    dict_add_number(dict, "start", !(prop->tp_flags & TP_FLAG_CONT_PREV));
+    dict_add_number(dict, "end", !(prop->tp_flags & TP_FLAG_CONT_NEXT));
+    pt = text_prop_type_by_id(buf, prop->tp_type);
+    if (pt != NULL)
+    dict_add_string(dict, "type", pt->pt_name);
```
Comment (k-takata) [inline src/textprop.c]: Indent is still wrong here.

File: src/textprop.c [hunk 2]
```diff
@@ -536,6 +554,193 @@ f_prop_clear(typval_T *argvars, typval_T *rettv UNUSED)
...
+    }
+
+    lnum_start = lnum;
+
+    if (rettv_dict_alloc(rettv) == FAIL)
+	return;
+    
+    while (1)
+    {	
+        char_u	*text = ml_get_buf(buf, lnum, FALSE);
```
Comment (k-takata) [inline src/textprop.c]: Indent of this `while` block looks wrong.

File: src/textprop.c (hunk 1 above)
Comment (r-hackett) [reply src/textprop.c]: That's fixed.

File: src/textprop.c (hunk 2 above)
Comment (r-hackett) [reply src/textprop.c]: Recreated the indents for prop_find, they should all be consistent. Not sure what happened here but looks good now.

**Comment by r-hackett** [comment]
Right, I'm pretty confident that the indents are consistent now. In my eagerness to get the code written I neglected these important details. I appreciate your patience.

AppVeyor's failing on a test but it doesn't seem related to this PR. The text prop tests pass and no warnings or errors are given compiling the modified c files.
```
From test_python3.vim:
...
Executed Test_unicode() in   0.001362 seconds
...
Executed 13 tests in   0.089168 seconds
1 FAILED:
Found errors in Test_unicode():
Caught exception in Test_unicode(): Vim(set):E543: Not a valid codepage: encoding=euc-tw @ function RunTheTest[40]..Test_unicode, line 14
```

When/if you guys are happy with this should I squash the commits and add a proper commit message?

**Comment by brammool** [comment]
Ryan Hackett wrote:

> That should fix the indents. A strange thing is that in the [diff](https://github.com/vim/vim/pull/5421/files#diff-be67b582e26932bd9217fba84b6c2e88) for textprop.c on github the variables at the top of prop_find are still not aligned, but in the [file viewer](https://github.com/vim/vim/blob/c0d8eb1a1c20a1bc595abe2b2c815be2f7c422f2/src/textprop.c) (and in vim with now I've made sure modeline is set) they're aligned. 

The diff usually has indents look wrong, because it inserts a margin of
one or two spaces before each line.

-- 
My girlfriend told me I should be more affectionate.
So I got TWO girlfriends.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by r-hackett** [comment]
> The diff usually has indents look wrong, because it inserts a margin of
> one or two spaces before each line.

Thanks for clearing that up. I'll keep an eye on this PR for any updates on your end. As far as I can see the indents are acceptable now.
